File: javascript/balanced_binary_tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {
    return dc(root, 1).balanced;
};

function dc(root, depth) {
    if (root === null) {
        return {
            balanced: true,
            height: depth - 1,
        };
    }

    var left = dc(root.left, depth + 1);
    var right = dc(root.right, depth + 1);

    return {
        balanced: Math.abs(left.height - right.height) <= 1 && left.balanced && right.balanced,
        height: Math.max(left.height, right.height),
    };
}

File: ruby/balanced_binary_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Boolean}
def is_balanced(root)
  is_balanced_helper(root).last
end

def is_balanced_helper(root)
  return [0, true] if root.nil?

  max_depth_left, balanced_left = is_balanced_helper(root.left)
  max_depth_right, balanced_right = is_balanced_helper(root.right)

  [[max_depth_left, max_depth_right].max + 1, (max_depth_left - max_depth_right).abs <= 1 && balanced_left && balanced_right]
end
